src/Lecture4.hs

```diff
@@ -184,7 +223,31 @@ instance for the 'Stats' type itself.
 -}
 
 instance Semigroup Stats where
-
+  (<>) :: Stats -> Stats -> Stats
+  s1 <> s2 =
+    Stats
+      { statsTotalPositions = statsTotalPositions s1 <> statsTotalPositions s2,
+        statsTotalSum = statsTotalSum s1 <> statsTotalSum s2,
+        statsAbsoluteMax = statsAbsoluteMax s1 <> statsAbsoluteMax s2,
+        statsAbsoluteMin = statsAbsoluteMin s1 <> statsAbsoluteMin s2,
+        statsSellMax = appendMaybeMax (statsSellMax s1) (statsSellMax s2),
+        statsSellMin = appendMaybeMin (statsSellMin s1) (statsSellMin s2),
+        statsBuyMax = appendMaybeMax (statsBuyMax s1) (statsBuyMax s2),
+        statsBuyMin = appendMaybeMin (statsBuyMin s1) (statsBuyMin s2),
+        statsLongest = statsLongest s1 <> statsLongest s2
+      }
+
+appendMaybeMax :: Maybe (Max Int) -> Maybe (Max Int) -> Maybe (Max Int)
+appendMaybeMax Nothing x = x
+appendMaybeMax x Nothing = x
+appendMaybeMax jmx@(Just (Max x)) jmy@(Just (Max y)) =
+  if x >= y then jmx else jmy
+
+appendMaybeMin :: Maybe (Min Int) -> Maybe (Min Int) -> Maybe (Min Int)
+appendMaybeMin Nothing x = x
+appendMaybeMin x Nothing = x
+appendMaybeMin jmx@(Just (Min x)) jmy@(Just (Min y)) =
+  if x <= y then jmx else jmy
```

Comment:
That's absolutely correct implementation 👍🏻 
You can write it a bit shorter though by implementing a single polymorphic function that pattern matches only on the final result:

```haskell
appendMaybeStrict :: Semigroup a => Maybe a -> Maybe a -> Maybe a
appendMaybeStrict a b = case a <> b of
    Nothing -> Nothing
    Just !x -> Just x
```